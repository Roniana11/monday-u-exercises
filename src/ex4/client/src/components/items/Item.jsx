import { ListItem, IconButton, Flex } from 'monday-ui-react-core';
import { ReactComponent as Delete } from '../../images/delete_icon.svg';
import { ReactComponent as Edit } from '../../images/edit_icon.svg';
import 'monday-ui-react-core/dist/main.css';
import { useRef, useState } from 'react';
import { PropTypes } from 'prop-types';
import { useDispatch } from 'react-redux';
import {
  editItemText,
  changeItemStatus,
  deleteItem,
} from '../../reducers/items-entities-reducer';

const Item = ({ text, id, status }) => {
  const dispatch = useDispatch();
  const label = useRef('');
  const [checked, setChecked] = useState(status);

  const deleteHandler = async () => {
    dispatch(deleteItem(id));
  };

  const changeStatusHandler = () => {
    dispatch(changeItemStatus(id, !checked));
    setChecked(!checked);
  };

  const editItemHandler = (event) => {
    const itemLabel = label.current;
    if (event.key === 'Enter') {
      itemLabel.contentEditable = 'false';
      if (itemLabel.prevtext !== itemLabel.innerText) {
        dispatch(editItemText(id, itemLabel.innerText));
      }
    }
  };

  const toggleEditMode = () => {
    const itemLabel = label.current;
    itemLabel.prevtext = itemLabel.innerText;
    const currentState = itemLabel.contentEditable;
    itemLabel.contentEditable = currentState === 'true' ? 'false' : 'true';
  };

  return (
    <ListItem>
      <Flex
        justify={Flex.justify.SPACE_BETWEEN}
        style={{
          width: '100%',
        }}
      >
        <div>
          <input
            type="checkbox"
            checked={checked}
            onChange={changeStatusHandler}
          ></input>
          <label ref={label} prevtext={text} onKeyPress={editItemHandler}>
            {text}
          </label>
        </div>
        <div>
          <IconButton
            ariaLabel="Edit item"
            icon={Edit}
            onClick={toggleEditMode}
          ></IconButton>
          <IconButton
            ariaLabel="Delete item"
            icon={Delete}
            onClick={deleteHandler}
          ></IconButton>
        </div>
      </Flex>
    </ListItem>
  );
};

export default Item;

Item.propTypes = {
  text: PropTypes.string,
  id: PropTypes.string,
  status: PropTypes.bool,
  onDelete: PropTypes.func,
  onEdit: PropTypes.func,
  onChangeStatus: PropTypes.func,
};
