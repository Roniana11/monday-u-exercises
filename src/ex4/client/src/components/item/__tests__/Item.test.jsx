import renderer from 'react-test-renderer';
import Item from '../Item';
import { Provider } from 'react-redux';
import { store } from '../../../store';

test('renders correctly an item', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Item id="1" text="test task" status={false} />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
