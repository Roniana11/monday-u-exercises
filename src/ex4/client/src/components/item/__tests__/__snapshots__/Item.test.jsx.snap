// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly an item 1`] = `
<div
  aria-disabled={false}
  className="list-item list-item--small"
  id=""
  onClick={[Function]}
  onKeyDown={[Function]}
  role="listitem"
  tabIndex={0}
>
  <div
    className="components-Flex-Flex-module__container--ZDN6j components-Flex-Flex-module__justifySpaceBetween--lQ3bb components-Flex-Flex-module__alignCenter--oWj46"
    style={
      Object {
        "gap": "0px",
        "width": "100%",
      }
    }
  >
    <div>
      <input
        checked={false}
        onChange={[Function]}
        type="checkbox"
      />
      <label
        onKeyPress={[Function]}
        prevtext="test task"
      >
        test task
      </label>
    </div>
    <div>
      <span
        className="components-IconButton-IconButton-modules__referenceContainer--BRxAY"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <button
          aria-disabled={false}
          aria-label="Edit item"
          className="monday-style-button monday-style-button--size-medium monday-style-button--kind-tertiary monday-style-button--color-primary monday-style-button--no-side-padding"
          data-testid="icon-button"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          style={
            Object {
              "alignItems": "center",
              "height": "40px",
              "justifyContent": "center",
              "padding": 0,
              "width": "40px",
            }
          }
          type="button"
        >
          <svg
            aria-hidden={false}
            className="icon_component icon-button-padding icon_component--clickable icon_component--no-focus-style"
            fill="currentColor"
            height="24"
            onClick={[Function]}
            role="button"
            tabIndex={0}
            viewBox="0 0 20 20"
            width="24"
          >
            <path
              clipRule="evenodd"
              d="M13.8542 3.59561C13.8541 3.59568 13.8542 3.59555 13.8542 3.59561L4.80915 12.6503L3.81363 16.189L7.35682 15.1957L16.4018 6.14C16.4746 6.06722 16.5161 5.96795 16.5161 5.86503C16.5161 5.76221 16.4753 5.6636 16.4026 5.59083C16.4025 5.59076 16.4026 5.59091 16.4026 5.59083L14.4038 3.59568C14.3309 3.52292 14.232 3.48197 14.1289 3.48197C14.026 3.48197 13.927 3.52297 13.8542 3.59561ZM12.8051 2.54754C13.1562 2.19695 13.6324 2 14.1289 2C14.6254 2 15.1016 2.19693 15.4527 2.54747C15.4527 2.5475 15.4527 2.54745 15.4527 2.54747L17.4515 4.54263C17.8026 4.89333 18 5.36914 18 5.86503C18 6.36091 17.8028 6.8365 17.4518 7.18719L8.26993 16.3799C8.17984 16.4701 8.06798 16.5356 7.94516 16.57L2.94244 17.9724C2.68418 18.0448 2.4069 17.9723 2.21725 17.7829C2.0276 17.5934 1.95512 17.3165 2.02768 17.0586L3.43296 12.0633C3.46728 11.9413 3.53237 11.8301 3.62199 11.7404L12.8051 2.54754Z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </button>
      </span>
      <span
        className="components-IconButton-IconButton-modules__referenceContainer--BRxAY"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <button
          aria-disabled={false}
          aria-label="Delete item"
          className="monday-style-button monday-style-button--size-medium monday-style-button--kind-tertiary monday-style-button--color-primary monday-style-button--no-side-padding"
          data-testid="icon-button"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          style={
            Object {
              "alignItems": "center",
              "height": "40px",
              "justifyContent": "center",
              "padding": 0,
              "width": "40px",
            }
          }
          type="button"
        >
          <svg
            aria-hidden={false}
            className="icon_component icon-button-padding icon_component--clickable icon_component--no-focus-style"
            fill="currentColor"
            height="24"
            onClick={[Function]}
            role="button"
            tabIndex={0}
            viewBox="0 0 20 20"
            width="24"
          >
            <path
              clipRule="evenodd"
              d="M8.30035 1.86462C7.77994 1.86462 7.29477 2.08976 6.94732 2.46719C6.60179 2.84253 6.41724 3.33927 6.41724 3.84552V4.32642H4.901H2.63477C2.22055 4.32642 1.88477 4.6622 1.88477 5.07642C1.88477 5.49063 2.22055 5.82642 2.63477 5.82642H4.151V16.1545C4.151 16.6608 4.33556 17.1575 4.68109 17.5328C5.02853 17.9103 5.51371 18.1354 6.03411 18.1354H13.9659C14.4863 18.1354 14.9715 17.9103 15.3189 17.5328C15.6645 17.1575 15.849 16.6608 15.849 16.1545V5.82642H17.3652C17.7794 5.82642 18.1152 5.49063 18.1152 5.07642C18.1152 4.6622 17.7794 4.32642 17.3652 4.32642H15.099H13.5828V3.84552C13.5828 3.33927 13.3982 2.84253 13.0527 2.46719C12.7053 2.08976 12.2201 1.86462 11.6997 1.86462H8.30035ZM7.16447 5.82642C7.16539 5.82642 7.16631 5.82642 7.16724 5.82642H12.8328C12.8337 5.82642 12.8346 5.82642 12.8356 5.82642H14.349V16.1545C14.349 16.3012 14.2948 16.4306 14.2153 16.5169C14.1378 16.6012 14.0465 16.6354 13.9659 16.6354H6.03411C5.95348 16.6354 5.86223 16.6012 5.78468 16.5169C5.7052 16.4306 5.651 16.3012 5.651 16.1545V5.82642H7.16447ZM12.0828 4.32642V3.84552C12.0828 3.69887 12.0286 3.56943 11.9491 3.4831C11.8716 3.39886 11.7803 3.36462 11.6997 3.36462H8.30035C8.21972 3.36462 8.12847 3.39886 8.05091 3.4831C7.97144 3.56943 7.91724 3.69887 7.91724 3.84552V4.32642L12.0828 4.32642Z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </button>
      </span>
    </div>
  </div>
</div>
`;
